In [10]:
import numpy as np

### Solución del problema:
### 1.a

Se genero las ecuaciones

In [16]:
A_1 = np.array([[2, 4, 5], 
                [6, 9, 8], 
                [4.1, 5, 3]])
B_1 = np.array([220, 490, 274])

Se le saca la inversa A y luego se multiplica por la misma matriz A, para sacar la matriz de identidad

In [17]:
# Get the inverse of A
A_inv_1 = np.linalg.inv(A_1)

# Multiply the inverse of A by A
result = np.dot(A_inv_1, A_1)

# Print the result
print(result)

[[ 1.00000000e+00 -7.10542736e-15 -7.10542736e-15]
 [-9.70164120e-16  1.00000000e+00  1.24344979e-14]
 [ 4.03096360e-16 -8.88178420e-16  1.00000000e+00]]


Multiplico la inversa de la Matriz A, por el vector B (Igualdad)

In [18]:
X_1 = np.linalg.inv(A_1).dot(B_1)
X_1

array([40., 10., 20.])

Se optiene los precops por hora de los procesadores A, B y C

In [23]:
print("El precio por hora del procesador A es:", X_1[0])
print("El precio por hora del procesador B es:", X_1[1])
print("El precio por hora del procesador B es:", X_1[2])

El precio por hora del procesador A es: 39.99999999999967
El precio por hora del procesador B es: 10.0000000000007
El precio por hora del procesador B es: 20.000000000000057


### Solución del problema:
### 1.b

Se corrige el error generado por el estudiante en el sistema de ecuaciones

In [19]:
A_2 = np.array([[2, 4, 5], 
                [6, 9, 8], 
                [4.2, 5, 3]])
B_2 = np.array([220, 490, 274])
X_2 = np.linalg.inv(A_2).dot(B_2)

In [20]:
# Get the inverse of A
A_inv_2 = np.linalg.inv(A_2)

# Multiply the inverse of A by A
result = np.dot(A_inv_2, A_2)

# Print the result
print(result)

[[ 1.00000000e+00 -1.77635684e-15 -8.88178420e-15]
 [ 1.09314267e-16  1.00000000e+00  8.88178420e-15]
 [ 1.29810692e-15 -8.88178420e-16  1.00000000e+00]]


In [21]:
x_2 = np.linalg.inv(A_2).dot(B_2)
x_2

array([20.        , 31.53846154, 10.76923077])

In [24]:
print("El precio por hora del procesador A es:", x_2[0])
print("El precio por hora del procesador B es:", x_2[1])
print("El precio por hora del procesador C es:", x_2[2])

El precio por hora del procesador A es: 20.00000000000013
El precio por hora del procesador B es: 31.538461538461764
El precio por hora del procesador C es: 10.769230769230754


Ahora vamos a calcular el error relativo del precio por hora del procesador A
siendo  x=39.99999999999967 y y=20.00000000000013 los resultados de los sistemas de ecuaciones solucionados

In [25]:
def relative_error(x, y):
    return abs(x - y) / abs(y)

In [26]:
x = 39.99999999999967
y = 20.00000000000013
error = relative_error(x,y)
print(error)

0.9999999999999705


Ahora vamos a calcular el error relativo del precio por hora del procesador B
siendo  x=10.0000000000007 y y=31.538461538461764 los resultados de los sistemas de ecuaciones solucionados 

In [27]:
x = 10.0000000000007
y = 31.538461538461764
error = relative_error(x,y)
print(error)

0.6829268292682728


Ahora vamos a calcular el error relativo del precio por hora del procesador C
siendo  x=20.000000000000057 y y=10.769230769230754 los resultados de los sistemas de ecuaciones solucionados 

In [28]:
x = 20.000000000000057
y = 10.769230769230754
error = relative_error(x,y)
print(error)

0.8571428571428651


Dando respueta a la pregunta <span style="color:green"> Resuelva el sistema para ver como cambia el
valor de la hora en cada procesador, antes y después del error. Explique
si el cambio es grande o no y por qué se da este cambio. </span>

El cambio es grande, ya que el error relativo de cada una de las soluciones no esta dando muy alto, por lo general se busca minimizar el error relativo para obtener una medida o cálculo más precisa y confiable

Ahora dando solución a <span style="color:green"> Consulte una
cota superior para el error relativo en la solución de un sistema cuando
hay un cambio en los coeficientes de la matriz del sistema. </span>

Para calcular la cota superior superior para el error relativo seria:

(∥x∥ / ∥Δx∥) ≤ κ(A) * (∥ΔA∥ / ∥A∥)

siendo k(A) = condicion​

La condición de una matriz, también conocida como número de condición, es una medida que se utiliza en el campo del análisis numérico para evaluar cuán sensible es una matriz a pequeños cambios en sus elementos o en sus términos de entrada

Debemos de generar la condición de la matriz A y para esto debemos de hallar la norma de A, por la norma de la matriz inversa de A

||A|| * ||A^(-1)||

In [4]:
A1 = np.array([[2, 4, 5], 
                [6, 9, 8], 
                [4.1, 5, 3]])

A2 = np.array([[2, 4, 5], 
               [6, 9, 8], 
               [4.2, 5, 3]])

Ahora calcula la condición de la matiz A1

In [6]:
cond_A1 = np.linalg.cond(A1, p=np.inf) # Usamos la norma 2 o infinita de matrices

Ahora calculamos la diferencia relativa entre A1 y A2 sobre la norma de A1

In [7]:
delta = np.linalg.norm(A1 - A2, ord=np.inf) / np.linalg.norm(A1, ord=np.inf)

Ahora se calcula el limite superior para el error relativo de Epsilon (cota suprior)

In [8]:
epsilon = cond_A1 * delta

In [9]:
print("La condición de la matriz A1:", cond_A1)
print("La diferencia relativa entre A1 y A2:", delta)
print("El limite superior o cota superior del error relativo es:", epsilon )

La condición de la matriz A1: 766.0769230769237
La diferencia relativa entre A1 y A2: 0.004347826086956545
El limite superior del error relativo es: 3.3307692307692514


R/ Como se puede ver la cota superior dio como resultado 3.3307692307692514 es que la matriz es muy sensible a pequeños cambios

Ahora dando respuesta a <span style="color:green"> Luego
calcule una cota superior para el error relativo en el valor de la hora
en los procesadores. </span>

Dado un vector de referencia A y un vector de aproximación B, el error relativo se puede calcular de la siguiente manera:

error relativo= ∥a∥ / ∥a−b∥

Y para calcular la cota superior se puede utilizar la desigualdad triangular y las propiedades de la norma

La cota superior para el error relativo se puede calcular de la siguiente manera:

cota superior del error relativo=  ∥a−b∥ / ∥a∥ ≤ ∥a∥+∥b∥ / ∥a∥

In [30]:
Vector_A = np.array([39.99999999999967, 10.0000000000007, 20.000000000000057])
Vector_A

array([40., 10., 20.])

In [31]:
Vector_B = np.array([20.00000000000013, 31.538461538461764, 10.769230769230754])
Vector_B

array([20.        , 31.53846154, 10.76923077])

In [33]:
error_relativo= np.abs(Vector_A - Vector_B) / np.abs(Vector_A)
error_relativo

array([0.5       , 2.15384615, 0.46153846])

In [38]:
cota_superior_del_valor = np.linalg.norm(error_relativo, ord=2)
cota_superior_del_valor

2.2587764400038135

R/ Como se puede ver la cota superior dio como resultado 2.2587764400038135 es que vector es muy sensible a pequeños cambios